### ISSUE TITLE
wcmatch glob.globmatch not matching some curly brace patterns

### ISSUE THREAD

**Issue by berglh** [facelessuser/wcmatch#182]
I'm not sure if this is a bug or not.

## What?

I'm trying to use `wcmatch` to parse S3 object keys (URI strings) to match file path globs that are supported in Pyspark (Scala/Java).

Spark uses something similar to the [Java File Glob](https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileSystem.html#getPathMatcher(java.lang.String)) parser.  I want something to match paths in a similar way in Python.  As per [#39856](https://bugs.python.org/issue39856) I can see that this is related to the curly brace issue in stdlib.

## Why?

Essentially, we're using Pyspark to process objects in AWS S3. The historical source objects (files) have many thousands of objects and take Spark quite some time to check to see if new files need processing against it's inbuilt checkpoint feature. We therefore limit what files can be read by the Spark reader using globs in the [Spark DataStreamReader](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.streaming.DataStreamReader.html#pyspark.sql.streaming.DataStreamReader) path. i.e.: `s3://bucket/prefix/20{2[0-2}}/*/*/*.parquet`

One of the problems with our streaming solution is that we want to monitor the Stream and shut it down after it's finished processing all files. Unfortunately, the Spark streaming object does not have anything to tell us what is different between a Stream starting and looking for new files, verse finished processing all new files and just waiting for new files to appear.  Therefore, we need to give Spark ample time to find all the matching files and start processing.  When Spark is actively processing, we can see from the stream status that it's actively processing files, and then when no more files are available we can stop.  But as I mentioned, we need to give it adequate time for at least starting processing before we can evaluate the `isDataAvailable` flag in the stream object to then shut it down.  We can estimate the time to wait based on the number of objects that Spark is trying to read.  We return all the S3 objects matching the prefix and want to use a glob parser to match the same subset used by the SparkReader.

## TLDR

Primarily, the curly braces expansion and the double star options are supported in Spark, but working glob patterns for the path matcher library don't seem to work in `wcmatch`.

Some working examples of Spark reader globs:
```python
from wcmatch import glob
path = "wal/prime/processed/2019/01/01/compilation_20190101_000.json.gz"
if glob.globmatch(path, 'wal/prime/processed/20{19,20,21,22}/*/*/*.json.gz', flags=glob.B):
    print(path)
```
> wal/prime/processed/2019/01/01/compilation_20190101_000.json.gz
```python
if glob.globmatch(path, 'wal/prime/processed/20{19,2[0-9]}/*/*/*.json.gz', flags=glob.B):
    print(path)
```
> wal/prime/processed/2019/01/01/compilation_20190101_000.json.gz

Here are a ~bunch~ couple of patterns that work in the Spark reader, but didn't in `wcmatch`:

**Edit**: I see I need to use `**/*.json.gz` in this case.
~~# braces and double star
if glob.globmatch(path, 'wal/prime/processed/20{19,20,21,22}/**.json.gz', flags=glob.B|glob.G):
    print(path)
if glob.globmatch(path, 'wal/prime/processed/20{19,20,21,22}/**.json.gz', flags=glob.G|glob.B):
    print(path)~~

```python
# nested braces
if glob.globmatch(path, 'wal/prime/processed/2{0{19,20,21,22}}/*/*/*.json.gz', flags=glob.B):
    print(path)
if glob.globmatch(path, 'wal/prime/processed/2{0{19,20,21,22}}/*/*/*.json.gz', flags=glob.B|glob.B):
    print(path)
```
**Edit**: Same as previous edit.
~~# double star
if glob.globmatch(path, 'wal/prime/processed/2019/**.json.gz', flags=glob.G):
    print(path)~~

Not sure if this is outside the scope of this project.

Thanks

**Comment by facelessuser** [comment]
I was just going to respond about the misuse of `**`, seems you solved that issue.

**Comment by facelessuser** [comment]
What is it you think the the nested braces should do?

**Comment by facelessuser** [comment]
Take a look at what Bash does with nested braces. Do you think this should match your pattern?

```
bash-5.1$ echo {0{1,2,3}}
{01} {02} {03}
```

**Comment by facelessuser** [comment]
I'm going to mark this as a `wontfix` as it appears wcmatch is working as expected. Braces are [rigorously tested](https://github.com/facelessuser/bracex/blob/master/tests/brace-cases.txt) against Bash, both nested and non-nested. I think the problem is the interpretation of what braces "should" be doing. In your examples, I don't understand why you are using the nested braces. If there is no sequence of options (comma separated values) in braces, the braces are treated as literal braces, just like bash.

It seems that there is some assumption that the braces should just be ignored? I'm not sure why that is though. It seems that you want to remove the outer braces. Then you should get what you want.

```
bash-5.1$ echo {0{1,2,3}}
{01} {02} {03}
bash-5.1$ echo 0{1,2,3}                                                                                                                      
01 02 03
```

If I'm misunderstanding something, please do correct me.

@gir-bot remove S: triage
@gir-bot add S: wontfix

**Comment by berglh** [comment]
@facelessuser 

I wash just trying this in bash as well.  There are other instances I can provide that might make this clearer for you.  Namely, we have some data sources with folder names in hours.  As we're in UTC+10, if we needed to replace a set of files from filed in UTC, we might need a sophisticated glob like this:
```
path = "wal/prime/processed/2019/12/31/23/compilation_20191231_2302.json.gz"
pattern = 'wal/prime/processed/{2019/12/31/{1[4-9],2[0-3]},2020/01/01/{0[0-9],1[0-3]}}/*.json.gz'
```
This definitely works in Spark, but doesn't match here. I'm guessing that double curly brace expansion in bash doesn't work this way, so it would make sense that this wouldn't be targeted. All I can say, is that it does work in Spark.

Thanks for your help :)

**Comment by berglh** [comment]
Actually, this is working, missed my glob flag. Closing.

**Comment by facelessuser** [comment]
I like it when bugs resolve themselves 🙂. 

The more complex the pattern, the more chances of user errors sneaking in. It's the same with regex.

I'm glad you figured it out. Don't hesitate to file an issue if you think you have found something.

> My guess is here that Spark allows the double curly braces when it doesn't expand the same in bash if there is no reason to do so. It's just a minor implementation thing, if we stick to the bash/Linux implementation in our references, we can still do what we want but just need to be mindful not to use bash incompatible globs.

We needed a reference, one that hopefully others are familiar with so there is less confusion. Bash was our choice. The thing is, not everyone knows all of these corner cases, so they still get confused anyways 🤷🏻 . It's also more confusing when other implementations take different approaches. Unfortunately, we can't match every implementation, but it sounds like you have a grasp of the situation now.

**Comment by berglh** [comment]
Yeah, I think this becomes a debate verse exact spec targeting verse erring on the side of compatibility. Never the less, me using this to match S3 URIs is a bit of a hack anyway, so if we are able to get the same functionality with a compatible pattern, there is nothing to fix.

**Comment by berglh** [comment]
I guess I should be petitioning to Spark to get them to add some initial start-up phase into the streaming object initialisation or actually provide more observability into the jobs that have been completed in the Spark session via the Python API.  Currently, they do not expose the file scan jobs, only jobs that process files. If we could see and count the file scan jobs vs the processing file jobs, then we could cancel the job without having to rely on some dynamic timing rule. We are trying to run lean shop here, so spawning Spark streaming sessions in k8s and tearing them down when idle saves us quite a bit of dough. This glob expansion will be a stop gap in the mean time for us in the meantime until we come up with something better in Spark.
